LoginServer/PipeServer.cs

```diff
@@ -0,0 +1,548 @@
+
+        /// <summary>
+        /// Event is called when ever a message is recieved from client pipe event.
+        /// </summary>
+        public event MessageReceivedHandler MessageRecieved;
+
+        /// <summary>
+        /// Handles the client disconnected massages.
+        /// </summary>
+        public delegate void ClientDisconnectedHandler();
+
+        /// <summary>
+        /// Event is called when a client pipe is severed.
+        /// </summary>
+        public event ClientDisconnectedHandler ClientDisconnected;
+
+        /// <summary>
+        /// Handles the get admin details method call.
+        /// </summary>
+        /// <returns></returns>
+        public delegate string GetAdminDetailsHandler();
+
+        /// <summary>
+        /// Event is called when the GetAdminDetails method is called.
+        /// </summary>
+        public event GetAdminDetailsHandler GetAdminDetails;
+
+        /// <summary>
+        /// Handles the AllowMessaging method calls.
+        /// </summary>
+        public delegate void AllowMessagingHandler();
+
+        /// <summary>
+        /// Event is called when the client has been validated.
+        /// </summary>
+        public event AllowMessagingHandler AllowMessaging;
+
+        #endregion
+
+        /// <summary>
+        /// Starts the server.
+        /// </summary>
+        /// <param name="pipename">The name of the pipe being used.</param>
+        public void Start(string pipename)
+        {
+            PipeNameIs(pipename);
+
+            listenThread = new Thread(ListenForClients)
+            {
+                IsBackground = true
+            };
+
+            listenThread.Start();
+
+            IsRunning(true);
+
+
+        }
+
+        /// <summary>
+        /// When a new client connects, sets the stream and creates a new thread to listen to them.
+        /// </summary>
+        public void ListenForClients()
+        {
+            SECURITY_DESCRIPTOR sd = new SECURITY_DESCRIPTOR();
+
+            // Set the Security Descriptor to be completely permissive. ////////////////////////////////////////////////
+            InitializeSecurityDescriptor(ref sd, SECURITY_DESCRIPTOR_REVISION);
+            SetSecurityDescriptorDacl(ref sd, true, IntPtr.Zero, false);
+
+            IntPtr ptrSD = Marshal.AllocCoTaskMem(Marshal.SizeOf(sd));
+            Marshal.StructureToPtr(sd, ptrSD, false);
+
+            SECURITY_ATTRIBUTES sa = new SECURITY_ATTRIBUTES
+            {
+                nLength = Marshal.SizeOf(sd),
+                lpSecurityDescriptor = ptrSD,
+                bInheritHandle = 1
+            };
+
+            IntPtr ptrSA = Marshal.AllocCoTaskMem(Marshal.SizeOf(sa));
+            Marshal.StructureToPtr(sa, ptrSA, false);
+
+            while (true)
+            {
+                // Create an instance of a named pipe for one client.
+                SafeFileHandle clientHandle = CreateNamedPipe(
```

Comment:
Create a new named pipe for every client?

Reply:
Or is this whole function only called when a client connects?
in which case, this is a loop because creating a pipe often fails?

This function confuses me. just becauase I cannot decide if its looping forever, creating new pipes,
That seems bad.
if its supposed to break out of the loop.  It has no break; statement in it.
If its supposed to be ccalled only on each connection, then why the loop?

very confusing.

Reply:
Right.

Your running this in a thread. and telling the thread to Abort() to stop the while loop running.
eh. it will work.  gotcha.